tests/conftest.py

```diff
@@ -170,13 +178,39 @@ def _generate_calculation_node(process_state=ProcessState.FINISHED, exit_status=
         if process_state is ProcessState.FINISHED and exit_status is None:
             exit_status = 0
 
-        node = CalculationNode(process_type=entry_point)
-        node.set_process_state(process_state)
+        calculation_node = CalculationNode(process_type=entry_point)
+        calculation_node.set_process_state(process_state)
 
         if exit_status is not None:
-            node.set_exit_status(exit_status)
+            calculation_node.set_exit_status(exit_status)
+
+        if repository is not None:
+            # Or use from_tree?
+            calculation_node.base.repository.put_object_from_filelike(repository[0], repository[1])
+
+        # For storing, need to first store the input nodes, then the CalculationNode, then the output nodes
+        if inputs is not None:
+            for input_label, input_node in inputs.items():
+                calculation_node.base.links.add_incoming(
+                    input_node,
+                    link_type=LinkType.INPUT_CALC,
+                    link_label=input_label,
+                )
+
+                input_node.store()
+
+        if outputs is not None:
+            # Need to first store CalculationNode before I can attach `created` outputs
+            calculation_node.store()
+            for output_label, output_node in outputs.items():
+                output_node.base.links.add_incoming(
+                    calculation_node, link_type=LinkType.CREATE, link_label=output_label
+                )
 
-        return node
+                output_node.store()
+
+        # Return unstored by default
+        return calculation_node
```

Comment:
Are any tests relying on the node returned being unstored?

Reply:
If the `calculation_node` is stored, this part here:
```python
        # Add sub-workchain that calls a calculation
        wc_node_sub.base.links.add_incoming(wc_node, link_type=LinkType.CALL_WORK, link_label='sub_workflow')
        for cj_node in cj_nodes:
            cj_node.base.links.add_incoming(wc_node_sub, link_type=LinkType.CALL_CALC, link_label='calculation')
```
https://github.com/qiaojunfeng/aiida-core/blob/9c151d7e4d453413350ca025e4c191ff1800e078/tests/tools/dumping/test_processes.py#L107-L110

leads to:
```python
        if self._node.is_stored:
            raise ValueError('attempted to add an input link after the process node was already stored.')
````
https://github.com/qiaojunfeng/aiida-core/blob/b7e59a0dbc0dd629be5c8178e98c70e7a2c116e9/src/aiida/orm/nodes/process/process.py#L101-L102

The whole business of storing/not storing nodes while generating the nested workchain to test the dumping seems very fragile. In its current state, it works. But storing nodes in other places, like the `calculation_node` here breaks things, while not storing them in other places (where they are currently being stored) leads to things (like `calculation_node` outputs) not being dumped. That's why the logic currently is in place as it is.